File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    
    public int solution(int[] numbers, int target) {
        //완전탐색(깊이탐색)을 하고 타겟에 닿았을 때 깊이가 numbers.length와 같으면 카운트 ++
        
        DFS(numbers, 0, 0, target);
        
        return answer;
    }
    public void DFS(int[] numbers, int depth, int sum, int target) {
        if(depth == numbers.length) {
            if(sum == target) {
                answer++;
            }
            
        } else {
            DFS(numbers, depth + 1, sum + numbers[depth], target);
            DFS(numbers, depth + 1, sum - numbers[depth], target);
            
        }
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# 탐색문제
# 하나의 노드당 +, - 둘다 실행을 해봐야한다.
# 별도의 함수를 호출했을 경우 어떻게 결괏값을 넘겨받을 수 있을지
def dfs(data, answer, depth, result, target):
    if depth == len(data):
        if result == target:
            answer[0] += 1
    else:
        dfs(data, answer, depth+1, result+data[depth], target)
        dfs(data, answer, depth+1, result-data[depth], target)
    
def solution(numbers, target):
    answer = [0]
    depth = 0
    result = 0
    dfs(numbers, answer, depth, result, target)
    return answer[0]